### PR TITLE
feat: handle `Record<>` mapped type

### LINKED ISSUE #653: mapped type: `Record<'foo', string>`
I'm trying to use `Record<'id', string>;`.
I've seen https://github.com/lukeautry/tsoa/pull/505 and it seems that it's known that `Record<>` aren't supported, but I wanted to have some update on this.

## Sorting

* **I'm submitting a ...**
  - [ ] bug report
  - [x] feature request
  - [ ] support request

* I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

```ts
export interface TestModel {
  recordType?: Record<'id', string>;
}
```

```
There was a problem resolving type of 'Record<'id', string>'.
There was a problem resolving type of 'TestModel'.

/private/tmp/tsoa/src/metadataGeneration/typeResolver.ts:151
      const tsProperties = this.current.typeChecker.getPropertiesOfType(type).map(symbol => {
                                                                              ^
TypeError: Cannot read property '0' of undefined
    at /private/tmp/tsoa/src/metadataGeneration/typeResolver.ts:151
```

Thank you

### PR DESCRIPTION
**Pull request by Eywek**
### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](https://github.com/lukeautry/tsoa/tree/master/docs/CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/lukeautry/tsoa/pulls) for the same update/change?
* [x] Have you written unit tests?
* [ ] Have you written unit tests that cover the negative cases (i.e.: if bad data is submitted, does the library respond properly)?
* [x] This PR is associated with an existing issue?

**Closing issues**

Closes #653

### If this is a new feature submission:

* [x] Has the issue had a maintainer respond to the issue and clarify that the feature is something that aligns with the [goals](https://github.com/lukeautry/tsoa#goal) and [philosophy](https://github.com/lukeautry/tsoa#philosophy) of the project?

**Potential Problems With The Approach**

Some code is duplicated with jsdoc tags, but I haven't found a better way to handle this.

### CODE REVIEW

**Comment by Eywek** [comment]
@WoH did you have time to check this one?

**Comment by WoH** [comment]
Not yet, I'll try to take a look over the weekend.
I'm worried about Type Alias compatibility, maybe I can come up with a way to ensure there are no unintended side effects here while reusing the existing jsDoc parsing helpers if possible, which is why I'd like to ask for some more time to review this.

File: src/metadataGeneration/typeResolver.ts [hunk 1]
```diff
@@ -142,23 +142,70 @@ export class TypeResolver {
...
-      const properties = tsProperties.filter(isNotIgnored).map(property => {
-        if (ts.isPropertySignature(property)) {
-          return this.propertyFromSignature(property, mappedTypeNode.questionToken);
-        } else if (ts.isPropertyDeclaration(property) || ts.isParameter(property)) {
-          return this.propertyFromDeclaration(property, mappedTypeNode.questionToken);
-        } else {
-          throw new GenerateMetadataError(`could not resolve property of kind ${property.kind}. Please report this as an issue`, property);
+      const typeChecker = this.current.typeChecker;
+      const properties: Tsoa.Property[] = [];
+      for (const property of type.getProperties()) {
```
Comment (WoH) [inline src/metadataGeneration/typeResolver.ts:148]: We should already map to the valuedeclaration and filter here.

**Comment by WoH** [comment]
First observations: 
- `Record<string, T>` won't set `additionalProperties`
- `Record<'name', Alias> won't ref to the alias from the property but inline and lose info.

You're removing calls to `getPropertyFromSignature` and `getPropertyFromDeclaration` and duplicate a lot of work. If these methods are not sufficient to resolve, we should improve these.

**Comment by Eywek** [comment]
> You're removing calls to propertyFromSignature and propertyFromDeclaration and duplicate a lot of work. If these methods are not sufficient to resolve, we should improve these.

I've tried to remove some duplication and use these methods. If you have some advice to remove other duplications I'm listening
I'll look for broken cases

**Comment by WoH** [comment]
> EDIT: I don't understand what you mean by Record<string, T>

A `Record<string, T>` is equivalent to `{[name:string]: T}`.
Example: `Record<string, number>` is like `{[name: string]: number}`

**Comment by Eywek** [comment]
Hm right I see. I don't really know how to properly handle that case and I don't know if `Record<>` should be used for this type of things anyway.
If you have any idea of how we can implement it I will try but otherwise where do you want to document it? In the readme or in a new `docs/` file?

**Comment by WoH** [comment]
> If you have any idea of how we can implement it I will try but otherwise where do you want to document it? In the readme or in a new docs/ file?

I'm working on new docs for the 3.0 stable, if I don't remember to document this, just remind me again.
No action needed from your side here.

I'm trying to understand why you need to reimplement the jsDoc parsing. Is the virtual node creating issues with the existing jsdoc parsing methods?

**Comment by Eywek** [comment]
> I'm trying to understand why you need to reimplement the jsDoc parsing. Is the virtual node creating issues with the existing jsdoc parsing methods (getNodeFormat, getNodeDescription, getNodeExample, getJSDocComment(node, 'default'))?

For example, when I'm trying to use `isExistJSDocTag(typeNode, tag => tag.tagName.text === 'ignore')` to ignore, it doesn't work on the `hidden` property (on the `ErrorResponseModel` interface which is used with `Omit<>` later). Furthermore, I can't use `getNodeDescription`, `getNodeExample`, `getNodeFormat` and `getJSDocComment` because I don't have a `node`, [here](https://github.com/lukeautry/tsoa/pull/662/files#diff-04ed568b4a647b0bfd718a5afdb822fbR194) `declaration` (which is called `node` in your methods) is undefined.

**Comment by WoH** [comment]
Ok, thanks for clarifying. I see that there is is are `tags` present on the `property`-Symbol, but I don't see how we may use that to set JsDoc annotations. This path only gets triggered with 'record-foo' and 'record-bar' right now, potentially also with some indexed type (?). In the first case there should not be any useful tags present, in the latter, they may potentially even be counter intuitive.

Trying to get some JSDoc from the `propertyType` seems hard, so maybe we shouldn't set these at all for now.

**Comment by Eywek** [comment]
So you want to remove https://github.com/lukeautry/tsoa/pull/662/files#diff-04ed568b4a647b0bfd718a5afdb822fbR183-R192 and https://github.com/lukeautry/tsoa/pull/662/files#diff-04ed568b4a647b0bfd718a5afdb822fbR173-R174 right?

I haven't found a way to test this code anyway so delete it could be a good move

**Comment by WoH** [comment]
Correct

**Comment by Eywek** [comment]
It's removed, about the description part, currently it resolves the description of the `Record<>` type. I don't know if it's wanted or not

**Comment by WoH** [comment]
> It's removed, about the description part, currently it resolves the description of the `Record<>` type. I don't know if it's wanted or not

```js
               description: 'Construct a type with a set of properties K of type T',
```

should come from the type alias description which is the of the property description, right?
I'm talking about the 2x `description: undefined` as the property descriptions.

**Comment by Eywek** [comment]
Hm, right

**Comment by WoH** [comment]
I wouldn't be sure where/when/how I'd want to set that, if there is a way to actually do that which makes sense, I'd be open, but currently, I don't think we actually need that (at least I can't tell)

**Comment by Eywek** [comment]
It should be good

**Comment by WoH** [comment]
LGTM, thank you so much for contributing and your patience!

**Comment by Eywek** [comment]
When do you plan to release?

**Comment by WoH** [comment]
Hope I'll be ready until tomorrow evening if that sounds good.

**Pull request merged**